### ISSUE TITLE
odd execution performance behavior

### ISSUE THREAD

**Issue by tve** [anywhichway/rule-reactor#10]
I have two seemingly simple rules that exhibit quite different behavior and I don't understand why. The first rule is:
```
reactor.createRule("is-public", 0, {r: Rule},
        function(r) {
                return r.fields.direction == "ingress" && r.fields.source_type == "cidr" &&
                        (r.fields.protocol == "tcp" || r.fields.protocol == "udp") &&
                        r.fields.cidr_ips.length > 4 &&
                        (r.fields.cidr_ips.startsWith("0.0.0.0") ||
                         r.fields.cidr_ips.charAt(r.fields.cidr_ips.length-2) == "/"); // /0 thru /9
        },
        function(r) {
                r.is_public = true;
                //console.log("rule", r.href, "is public:", r.fields.cidr_ips);
        });
```
When run on a fact set consisting of 5 facts I get the following stats:
```
is-public -- matches:5 tested:1 activated:3 fired:3
```
However, the following rule, which seems very similar to me gets very different stats:
```
reactor.createRule("set-security-group", 0, {r: Rule},
        function(r) { return r.fields.links != null && !r.security_group_href; },
        function(r) {
                r.security_group_href = null;
                for (i=0; !r.security_group_href && i<r.fields.links.length; i++) {
                        if (r.fields.links[i].rel == "security_group")
                                r.security_group_href = r.fields.links[i].href;
                }
                //console.log("set rule's security group href to", r.security_group_href);
        });
```
Stats:
```
set-security-group -- matches:5 tested:11 activated:10 fired:5
```
In particular, there is a huge performance difference in that the first rule is tested only once while the second is tested 11 times. A real test of this inserts several thousand facts and the difference then is 1 vs thousands. (I realize that the one test really is a once over the array.) Can you explain why the two rules behave so differently?

Here is a run on a set of 1381 rules:
```
> time node openports1.js
Asserted 1381 rules
Data Count:  1381
Executions:  1060
RPS:  66250
is-public -- matches:1381 tested:1 activated:1059 fired:1059
stop -- matches:0 tested:1 activated:1 fired:1
0.29user 0.01system 0:00.32elapsed 95%CPU (0avgtext+0avgdata 39988maxresident)k
1376inputs+0outputs (0major+6082minor)pagefaults 0swaps
> time node openports2.js
Asserted 1381 rules
Data Count:  1381
Executions:  1382
RPS:  648.8262910798121
set-security-group -- matches:1381 tested:2763 activated:2762 fired:1381
stop -- matches:0 tested:1 activated:1 fired:1
0.63user 0.04system 0:02.41elapsed 27%CPU (0avgtext+0avgdata 48680maxresident)k
0inputs+0outputs (0major+8325minor)pagefaults 0swaps
```
Note the significant performance difference: 0.32s vs. 2.41s

I'm attaching the two test files I used for the small 5-rule examples. I run them using node.js, e.g. simply `node openports1.js`.

[examples.zip](https://github.com/anywhichway/rule-reactor/files/480761/examples.zip)

Thanks much!

**Comment by tve** [comment]
After some more digging, I gather the above performance discrepancy comes from the fact that the condition of the set-security-group rule invokes the getter of r.security_group_href and its action invokes the setter?

The following simpler test still has me stumped:
```
//var reactor = new RuleReactor({});
var RuleReactor = require("rule-reactor");
var reactor = new RuleReactor();

function Obj(r) { Object.assign(this, r); }
global.Obj = Obj;

reactor.createRule("add field", 0, {o: Obj},
        function(o) { return (o.id & 1) == 0; },
        function(o) { o.even = true; }
);

reactor.createRule("stop", -100, {},
    function() { return true; },
    function() { reactor.stop(); }
);

// define objects
for (var i=0; i<100; i++) {
        reactor.assert(new Obj({name: "object "+i, id: i, odd: null}));
}

reactor.run(Infinity, false, function() {
        console.log("Data Count: ", reactor.data.size);
        console.log("Executions: ", reactor.run.executions);
        console.log("RPS: ", reactor.run.rps);
        Object.keys(reactor.rules).forEach(function(rulename) {
                var rule = reactor.rules[rulename];
                console.log(rule.name, "-- matches:"+rule.potentialMatches,
                        "tested:"+rule.tested, "activated:"+rule.activated, "fired:"+rule.fired);
        });
});
```
Results in:
```
> node test1.js
Data Count:  100
Executions:  2551
RPS:  54276.59574468085
add field -- matches:100 tested:100 activated:2550 fired:2550
stop -- matches:0 tested:1 activated:1 fired:1
```
Why is this rule fired 2550 times? And how would you write it better?

**Comment by tve** [comment]
I did some more tests. In the above example, I added some printing to the "add field" rule:
```
LL = console.log;

reactor.createRule("add field", 0, {o: Obj},
        function(o) { LL("add?", o.id); return (o.id & 1) == 0; },
        function(o) { LL("add!", o.id); o.even = true; }
);
```
and i changed the loop that creates the objects to only create 4 instead of 100. The execution log I get is:
```
> node test1.js
add? 0
add? 0
add? 1
add? 0
add? 1
add? 2
add? 0
add? 1
add? 2
add? 3
start
add? 2
add! 2
add? 0
add! 0
add? 2
add! 2
add? 0
add! 0
add? 0
add! 0
add? 0
add! 0
Data Count:  4
Executions:  7
RPS:  2333.3333333333335
add field -- matches:4 tested:4 activated:6 fired:6
stop -- matches:0 tested:1 activated:1 fired:1
```
- why is the condition re-evaluated on all facts each time I insert a new fact?
- why is the condition evaluated 3x in a row on the object with id==0 at the end of the run?
- what does "matches:4" and "tested:4" refer to?

**Comment by anywhichway** [comment]
We will did into this and get you an answer by next Monday, Sept 26th at the latest. Thanks for the simplified example.

**Comment by tve** [comment]
I think the following patch fixes the extraneous condition re-evaluations (the line numbers are off by a tad):
```
@@ -972,19 +995,15 @@ var uuid = require("uuid");
                        }
                });
                // test all associated rules
-               var rulestotest = {};
+               var rulestotest = [];
                instancestoprocess.forEach(function(instance) {
                        if(instance.rules) {
                                Object.keys(instance.rules).forEach(function(rulename) {
                                        var rule = instance.rules[rulename];
-                                       //promises.push(instance.rules[ruleinstance].test(instance));
-                                       rulestotest[rulename] = rule;
+                                       rule.test(instance);
                                });
                        }
                });
-               Object.keys(rulestotest).forEach(function(rulename) {
-                       rulestotest[rulename].test();
-               });
                if(callback) {
                        callback(null);
                }
```

**Comment by anywhichway** [comment]
Copying your comment from #11. "This test case is fixed by the patch to #10. I'm not sure that patch fixes all instances of this problem. The take-away is that rule.test must only be called on exactly the combinations that truly need to be retested, not fewer but also not any more. That seems fragile."

You are correct. A rule should only be retested on only the combinations that need to be re-tested. This  is what a pattern/join net maintains in a Rete or TREAT based system and one of the things that makes them so memory heavy. And, you obviously found the bug. Note, when we were using promises we called rule.test properly. When we stopped using promises we neglected to pass in the instance to test. Working to integrate this now.
